### PR TITLE
Allow more flexible CosmoPower network settings.

### PR DESCRIPTION
**Pull request by HTJense**
[CosmoPower](https://github.com/simonsobs/SOLikeT/blob/master/soliket/cosmopower.py) as it currently stands _requires_ the user to include a TT/TE/EE network of NN/PCAplusNN/NN type. Since this is a pretty strict set of requirements, this commit works on changing that.

Instead of each network being defined only by a network path, there is now a `network_settings` dict within the CosmoPower theory block, which lists each network and what type of network it specifies. This allows one to:

- include more or fewer networks than just TT/TE/EE (think inclusion of BB or pp).
- Added support for pp.
- specify the type of network for each spectrum. No longer is TE specified to be a PCA+NN type, but one can specify the network type for each network individually.
- The required input parameters are now checked at initialization instead.
- (4400ef093f89238cf073516aa8a43d03615bd684) If your network has been trained on D_ell instead of C_ell, you can set a flag that will take care of the prefactor (a necessity, based on personal experience).
- (cdf3e806d1877b22a9dc516082b956873a10e82f) Added networks for derived parameter calculation.

Some things that might be worthwhile adding before or after this merge include:
- Ensure that networks trained over different ell-ranges are supported (right now it still asserts every network operates over the same ell range, but this may not be necessarily true, see also point below).
- P(k), see also #84.
- ~~Derived parameter networks (see Boris Bolliet's work).~~ Added in cdf3e806d1877b22a9dc516082b956873a10e82f.
- A better pytest implementation.

### CODE REVIEW

**Comment by HTJense** [comment]
Current state is good enough to go into main branch for me, it already allows for an improved experience by end-users.
Extra features (i.e. non-Cl spectra, derived parameters, better error coverage, etc.) is probably more fitting for a different pull request, to prevent lengthy review & development hell.

File: soliket/cosmopower.py [hunk 1]
```diff
@@ -69,38 +90,82 @@ def calculate(self, state, want_derived=True, **params):
...
+            cls[k][:2] = 0.0
             if np.any(np.isnan(cls[k])):
                 self.log.warning("CosmoPower used outside of trained "\
                                  "{} ell range. Filled in with NaNs.".format(k))
 
         return cls
 
-    def get_can_support_params(self):
-        return ["omega_b", "omega_cdm", "h", "logA", "ns", "tau_reio"]
+    def ell_factor(self, ls: np.ndarray, spectra: str) -> np.ndarray:
```
Comment (itrharrison) [inline soliket/cosmopower.py:139]: Maybe mention somewhere (either here or in the test dict) that these ell_factor conversions specifically refer to CLASS units.

**Review by itrharrison — CHANGES_REQUESTED** [review]
Looks good, just need to mention that the ell_factors are for CLASS units.

File: soliket/cosmopower.py (hunk 1 above)
Comment (HTJense) [reply soliket/cosmopower.py:139]: They refer to the generic Cl/Dl prefactor, see for example [here](https://camb.readthedocs.io/en/latest/results.html#camb.results.CAMBdata.get_cmb_power_spectra). They are used to resolve the `ell_factor` option in `get_Cl` which is [defined in BoltzmannBase](https://github.com/CobayaSampler/cobaya/blob/master/cobaya/theories/cosmo/boltzmannbase.py#L312). (I have never used class so I don't refer to anything from class specifically).

**Comment by HTJense** [comment]
Any updates on this @itrharrison ?

**Review by itrharrison — DISMISSED** [review]
It would be good to have a more extensive test suite, but I am happy with this for now.

**Review by itrharrison — APPROVED** [review]

**Pull request merged**